From 508e0cc43be9b617cdd4ac5415c76521d478ef92 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 10 Jun 2007 16:10:17 +0000
Subject: [PATCH] CONFIGURATION-280: Fixed possible data loss for file-based
 configurations in auto-save mode that are associated with a reloading
 strategy; thanks to Roman Kurmanowytsch

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/configuration/trunk@545904 13f79535-47bb-0310-9956-ffa450edef68
---
 .../AbstractFileConfiguration.java            |  3 +
 .../configuration/TestXMLConfiguration.java   | 69 ++++++++++---------
 xdocs/changes.xml                             |  5 ++
 3 files changed, 46 insertions(+), 31 deletions(-)

diff --git a/src/java/org/apache/commons/configuration/AbstractFileConfiguration.java b/src/java/org/apache/commons/configuration/AbstractFileConfiguration.java
index af772611ba..6aa0f72f33 100644
--- a/src/java/org/apache/commons/configuration/AbstractFileConfiguration.java
+++ b/src/java/org/apache/commons/configuration/AbstractFileConfiguration.java
@@ -827,6 +827,8 @@ public void reload()
                         }
                         fireEvent(EVENT_RELOAD, null, getURL(), true);
                         setDetailEvents(false);
+                        boolean autoSaveBak = this.isAutoSave(); // save the current state
+                        this.setAutoSave(false); // deactivate autoSave to prevent information loss
                         try
                         {
                             clear();
@@ -834,6 +836,7 @@ public void reload()
                         }
                         finally
                         {
+                            this.setAutoSave(autoSaveBak); // set autoSave to previous value
                             setDetailEvents(true);
                         }
                         fireEvent(EVENT_RELOAD, null, getURL(), false);
diff --git a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
index 40c1d44022..9c165c9a27 100644
--- a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
@@ -263,7 +263,7 @@ public void testSetAttribute()
         // set a new attribute
         conf.setProperty("foo[@bar]", "value");
         assertEquals("foo[@bar]", "value", conf.getProperty("foo[@bar]"));
-        
+
         conf.setProperty("name1","value1");
         assertEquals("value1",conf.getProperty("name1"));
     }
@@ -352,11 +352,11 @@ public void testLoadAndSaveFromFile() throws Exception
         assertTrue(conf.isEmpty());
         conf.addProperty("test", "yes");
         conf.save();
-        
+
         conf = new XMLConfiguration(testSaveConf);
         assertEquals("yes", conf.getString("test"));
     }
-    
+
     /**
      * Tests loading a configuration from a URL.
      */
@@ -367,7 +367,7 @@ public void testLoadFromURL() throws Exception
         assertEquals("value", conf.getProperty("element"));
         assertEquals(url, conf.getURL());
     }
-    
+
     /**
      * Tests loading from a stream.
      */
@@ -377,12 +377,12 @@ public void testLoadFromStream() throws Exception
         conf = new XMLConfiguration();
         conf.load(new ByteArrayInputStream(xml.getBytes()));
         assertEquals(1, conf.getInt("test"));
-        
+
         conf = new XMLConfiguration();
         conf.load(new ByteArrayInputStream(xml.getBytes()), "UTF8");
         assertEquals(1, conf.getInt("test"));
     }
-    
+
     /**
      * Tests loading a non well formed XML from a string.
      */
@@ -439,7 +439,7 @@ public void testSave() throws Exception
         {
            conf.addProperty("test.attribute[@array]", "value" + i);
         }
-        
+
         // add comma delimited lists with escaped delimiters
         conf.addProperty("split.list5", "a\\,b\\,c");
         conf.setProperty("element3", "value\\,value1\\,value2");
@@ -453,7 +453,7 @@ public void testSave() throws Exception
         checkConfig.setFileName(testSaveConf.getAbsolutePath());
         checkSavedConfig(checkConfig);
     }
-    
+
     /**
      * Tests saving to a URL.
      */
@@ -464,7 +464,7 @@ public void testSaveToURL() throws Exception
         checkConfig.setFile(testSaveConf);
         checkSavedConfig(checkConfig);
     }
-    
+
     /**
      * Tests saving to a stream.
      */
@@ -485,11 +485,11 @@ public void testSaveToStream() throws Exception
                 out.close();
             }
         }
-        
+
         XMLConfiguration checkConfig = new XMLConfiguration();
         checkConfig.setFile(testSaveConf);
         checkSavedConfig(checkConfig);
-        
+
         try
         {
             out = new FileOutputStream(testSaveConf);
@@ -502,7 +502,7 @@ public void testSaveToStream() throws Exception
                 out.close();
             }
         }
-        
+
         checkConfig.clear();
         checkSavedConfig(checkConfig);
     }
@@ -518,13 +518,13 @@ public void testAutoSave() throws Exception
         // reload the configuration
         XMLConfiguration conf2 = new XMLConfiguration(conf.getFile());
         assertEquals("'autosave' property", "ok", conf2.getString("autosave"));
-        
+
         conf.clearTree("clear");
         conf2 = new XMLConfiguration(conf.getFile());
         Configuration sub = conf2.subset("clear");
         assertTrue(sub.isEmpty());
     }
-    
+
     /**
      * Tests if a second file can be appended to a first.
      */
@@ -536,14 +536,14 @@ public void testAppend() throws Exception
         conf.load(testProperties2);
         assertEquals("value", conf.getString("element"));
         assertEquals("tasks", conf.getString("table.name"));
-        
+
         conf.save(testSaveConf);
         conf = new XMLConfiguration(testSaveConf);
         assertEquals("value", conf.getString("element"));
         assertEquals("tasks", conf.getString("table.name"));
         assertEquals("application", conf.getString("table[@tableType]"));
     }
-    
+
     /**
      * Tests saving attributes (related to issue 34442).
      */
@@ -749,12 +749,12 @@ public void testValidating() throws ConfigurationException
         File nonValidFile = new File("conf/testValidateInvalid.xml");
         conf = new XMLConfiguration();
         assertFalse(conf.isValidating());
-        
+
         // Load a non valid XML document. Should work for isValidating() == false
         conf.load(nonValidFile);
         assertEquals("customers", conf.getString("table.name"));
         assertFalse(conf.containsKey("table.fields.field(1).type"));
-        
+
         // Now set the validating flag to true
         conf.setValidating(true);
         try
@@ -767,7 +767,7 @@ public void testValidating() throws ConfigurationException
             //ok
         }
     }
-    
+
     /**
      * Tests handling of empty elements.
      */
@@ -778,12 +778,12 @@ public void testEmptyElements() throws ConfigurationException
         conf.addProperty("empty2", "");
         conf.setProperty("empty", "no more empty");
         conf.save(testSaveConf);
-        
+
         conf = new XMLConfiguration(testSaveConf);
         assertEquals("no more empty", conf.getString("empty"));
         assertEquals("", conf.getProperty("empty2"));
     }
-    
+
     /**
      * Tests whether the encoding is correctly detected by the XML parser. This
      * is done by loading an XML file with the encoding "UTF-16". If this
@@ -798,7 +798,7 @@ public void testLoadWithEncoding() throws ConfigurationException
         conf.load(file);
         assertEquals("test3_yoge", conf.getString("yoge"));
     }
-    
+
     /**
      * Tests whether the encoding is written to the generated XML file.
      */
@@ -813,7 +813,7 @@ public void testSaveWithEncoding() throws ConfigurationException
         assertTrue("Encoding was not written to file", out.toString().indexOf(
                 "encoding=\"" + ENCODING + "\"") >= 0);
     }
-    
+
     /**
      * Tests whether a default encoding is used if no specific encoding is set.
      * According to the XSLT specification (http://www.w3.org/TR/xslt#output)
@@ -938,7 +938,7 @@ public void testXPathExpressionEngine()
         conf.clear();
         assertNull(conf.getString("test[1]/entity/@name"));
     }
-    
+
     /**
      * Tests the copy constructor.
      */
@@ -1131,6 +1131,19 @@ public void testMultipleAttrValuesEscaped() throws ConfigurationException
         checkSavedConfig(checkConf);
     }
 
+    /**
+     * Tests a combination of auto save = true and an associated reloading
+     * strategy.
+     */
+    public void testAutoSaveWithReloadingStrategy() throws ConfigurationException
+    {
+        conf.setFile(testSaveConf);
+        conf.save();
+        conf.setReloadingStrategy(new FileAlwaysReloadingStrategy());
+        conf.setAutoSave(true);
+        assertEquals("Value not found", "value", conf.getProperty("element"));
+    }
+
     /**
      * Prepares a configuration object for testing a reload operation.
      *
@@ -1168,12 +1181,6 @@ private void removeTestFile()
     private void checkSavedConfig(FileConfiguration checkConfig) throws ConfigurationException
     {
         checkConfig.load();
-
-        for (Iterator i = conf.getKeys(); i.hasNext();)
-        {
-            String key = (String) i.next();
-            assertTrue("The saved configuration doesn't contain the key '" + key + "'", checkConfig.containsKey(key));
-            assertEquals("Value of the '" + key + "' property", conf.getProperty(key), checkConfig.getProperty(key));
-        }
+        ConfigurationAssert.assertEquals(conf, checkConfig);
     }
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 73da724e10..907f57647c 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,11 @@
 
   <body>
     <release version="1.5-SNAPSHOT" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-280"
+        due-to="Roman Kurmanowytsch">
+        Using file-based configurations in auto-save mode together with a
+        reloading strategy could cause data loss. This has been fixed.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-279">
         A PropertiesConfiguration that was created from a non existing file
         lost its content when it was saved. This problem has been solved.
